judo/app/data/controller_data.py

```diff
@@ -0,0 +1,174 @@
+# Copyright (c) 2025 Robotics and AI Institute LLC. All rights reserved.
+# TODO(@bhung): We need to figure out how to properly test this.
+
+import time
+
+import numpy as np
+from omegaconf import DictConfig
+
+from judo.app.structs import MujocoState, SplineData
+from judo.app.utils import register_optimizers_from_cfg, register_tasks_from_cfg
+from judo.controller import Controller, ControllerConfig
+from judo.optimizers import Optimizer, OptimizerConfig, get_registered_optimizers
+from judo.tasks import Task, TaskConfig, get_registered_tasks
+
+
+class ControllerData:
+    """Base class that encompasses the minimal amount of data for a controller node to access.
+
+    This class is a small container which includes the data required for a node to run. This include configurations,
+    a controller, task information, and an optimizer. Middleware nodes should keep data within this class and
+    implement methods to send, process, and receive data from here.
+    """
+
+    def __init__(
+        self,
+        init_task: str,
+        init_optimizer: str,
+        task_registration_cfg: DictConfig | None = None,
+        optimizer_registration_cfg: DictConfig | None = None,
+    ) -> None:
+        """Initializes basic attributes of a controller data container."""
+        self.task_name = init_task
+        if task_registration_cfg is not None:
+            register_tasks_from_cfg(task_registration_cfg)
+        if optimizer_registration_cfg is not None:
+            register_optimizers_from_cfg(optimizer_registration_cfg)
+        self.last_plan_time = 0.0
+        self.paused = False
+        self.available_optimizers = get_registered_optimizers()
+        self.available_tasks = get_registered_tasks()
+        self._setup(init_task, init_optimizer)
+
+    def _setup(self, task_name: str, optimizer_name: str) -> None:
+        """Set up the task and optimizer for the controller."""
+        task_entry = self.available_tasks.get(task_name)
+        optimizer_entry = self.available_optimizers.get(optimizer_name)
+
+        assert task_entry is not None, f"Task {task_name} not found in task registry."
+        assert optimizer_entry is not None, f"Optimizer {optimizer_name} not found in optimizer registry."
+
+        # instantiate the task/optimizer/controller
+        task_cls, task_config_cls = task_entry
+        optimizer_cls, optimizer_config_cls = optimizer_entry
+
+        task = task_cls()
+        optimizer_config = optimizer_config_cls()
+        optimizer = optimizer_cls(optimizer_config, task.nu)
+
+        controller_config = ControllerConfig()
+        controller_config.set_override(task_name)
+        self.controller = Controller(
```

Comment:
Wait also this struct is called `ControllerData` but sits above the `Controller`? I am very confused